sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala

```diff
@@ -142,12 +143,14 @@ case class MapGroups(
  */
 case class CoGroup(
     func: (Any, Iterator[Any], Iterator[Any]) => TraversableOnce[Any],
-    keyObject: Expression,
-    leftObject: Expression,
-    rightObject: Expression,
+    keyDeserializer: Expression,
+    leftDeserializer: Expression,
+    rightDeserializer: Expression,
     serializer: Seq[NamedExpression],
     leftGroup: Seq[Attribute],
     rightGroup: Seq[Attribute],
+    leftAttr: Seq[Attribute],
+    rightAttr: Seq[Attribute],
```

Comment:
I might be missing something here, but it looks like there is only one place where we create a `CoGroup` operator and it is always using just the left/right output attributes.  Lets not over generalize unless it is useful.